### PR TITLE
Dropdown a11y updates

### LINKED ISSUE #1147: Dropdown needs accessibility improvements
The [Dropdown component](http://www.carbondesignsystem.com/components/dropdown/code) (I only tried the first one on the code page) has several accessibility problems:
1. Having a label that disappears is not considered very accessible (similar to [placeholder in text input not being considered very accessible](https://www.smashingmagazine.com/2018/06/placeholder-attribute/)). I have seen devs try to give this a label with `<label>` (which doesn't work here) or hook up a span with aria-labelledby. Fixing this "disappearing label" problem may involve a slight redesign.  :)
2. Typing Esc closes the list, but it also closes the side nav, so the key handler for Esc needs to stop propagation and maybe also prevent default.
3. The markup is confusing to screen readers:
   - it thinks this is a "list with 2 items": the bx--dropdown-text item and the item containing the dropdown list. It may help to to change the outer `<ul>` and the 2 `<li>` into divs (or give them role="presentation").
   - it doesn't know that the dropdown list drops down, so it doesn't tell the user that there's a dropdown available
   - it reads all of the items in the dropdown list right away, even though they are not showing

Some aria may be needed to make this more understandable (maybe aria-haspopup="listbox" and aria-expanded="true/false"). Please read the ["menu button" pattern](http://w3c.github.io/aria-practices/#menubutton) and the ["combobox" pattern](http://w3c.github.io/aria-practices/#combobox)  in the ARIA Authoring Practices Guide to see what makes the most sense. They are both similar to what you are trying to achieve, I think.

### PR DESCRIPTION
**Pull request by elizabethsjudd**
Closes: #1147

Update to Dropdown for a11y.

#### Changelog

**Changed**

- Dropdown HTML is updated for a11y considerations using semantic tags, aria attributes, and role attributes
- Added styles for the new structure
- JavaScript updated to use new structure and support `aria-activedescendant`

#### Testing / Reviewing

- regression test the dropdown for all users types (screen reader, keyboard, mouse)

### CODE REVIEW

**Comment by elizabethsjudd** [comment]
@carbon-design-system/developers is there a reason dropdown has the `data-floating-menu-direction="bottom"` attribute but doesn't use `FloatingMenu` in the scripts?

**Review by asudoh — COMMENTED** [review]
Looks mostly markup change so it looks good from dev perspective, adding @carbon-design-system/design to ensure the visual is still good.

**Comment by elizabethsjudd** [comment]
@asudoh can you answer this question: https://github.com/carbon-design-system/carbon/pull/3586#issuecomment-516589845? It won't be part of this PR but it impacts the stuff that I'm working on now for Dropdown menu

**Comment by laurenmrice** [comment]
Screen Shot 2019-07-31 at 10 31 02 AM


Is this focus on the whole drawer browser determined when it comes to the styling of it? Because it looks different from our regular focus @elizabethsjudd

**Comment by elizabethsjudd** [comment]
@laurenmrice 

> Right now default browser focus is visible on the menu but that will be corrected in the next PR.

the fix for that is going to come in the next PR. The whole menu technically gets focus but then we'll use the `aria-activedescendent` attribute to navigate the dropdown which will give the menu items the appearance of focus.  I'm almost done with the next PR (hoping EOD today or sometime tomorrow).

**Comment by elizabethsjudd** [comment]
@asudoh @laurenmrice The updates to the JavaScript wasn't that much so I combined my two branches in to this single PR.

**Comment by asudoh** [comment]
Quick answer on your floating menu question (sorry @elizabethsjudd for missing this!) - The attribute seems to have been added by mistake - Dropdown is _not_ a floating menu, and my quick look tells me that dropdown code/style (still) does _not_ refer to that attribute.

File: packages/components/src/components/dropdown/dropdown.js [hunk 1]
```diff
@@ -82,6 +88,9 @@ class Dropdown extends mixin(
    * Opens and closes the dropdown menu.
    * @param {Event} [event] The event triggering this method.
    */
+  // @question this function is called every time I click on the document and executes for
```
Comment (asudoh) [inline packages/components/src/components/dropdown/dropdown.js]: IIRC this is for click-outside or focus-outside. Given some of the logic is not relevant to that, we may split the logic, one for global document listener, another for root element listener.

File: packages/components/src/components/dropdown/dropdown.js [hunk 2]
```diff
@@ -15,6 +15,7 @@ import on from '../../globals/js/misc/on';
 
 const toArray = arrayLike => Array.prototype.slice.call(arrayLike);
 
+// @question does/should dropdown actually use Floating Menu?
```
Comment (asudoh) [inline packages/components/src/components/dropdown/dropdown.js]: No.

File: packages/components/src/components/dropdown/dropdown.js [hunk 3]
```diff
@@ -39,6 +40,11 @@ class Dropdown extends mixin(
    */
   constructor(element, options) {
     super(element, options);
+    this.triggerNode = this.element.querySelector(this.options.selectorTrigger);
```
Comment (asudoh) [inline packages/components/src/components/dropdown/dropdown.js]: Our codebase doesn't keep node refs as instance properties. Good to keep doing the same unless we make a decision to make the change for the entire codebase.

**Review by asudoh — COMMENTED** [review]
Good progress @elizabethsjudd!

**Review by laurenmrice — APPROVED** [review]

File: packages/components/src/components/dropdown/dropdown.js (hunk 3 above)
Comment (elizabethsjudd) [reply packages/components/src/components/dropdown/dropdown.js]: @asudoh Sorry I'm not sure what you are asking here. Are you saying i should query for the elements in each function that I use them?

File: packages/components/src/components/dropdown/dropdown.js (hunk 1 above)
Comment (elizabethsjudd) [reply packages/components/src/components/dropdown/dropdown.js]: @asudoh I feel like this is beyond the scope of this PR but do you want me to open up an issue to look in to refactoring this?

Comment (asudoh) [reply packages/components/src/components/dropdown/dropdown.js]: Sounds good to me - Thanks @elizabethsjudd!

File: packages/components/src/components/dropdown/dropdown.js (hunk 3 above)
Comment (asudoh) [reply packages/components/src/components/dropdown/dropdown.js]: You are right @elizabethsjudd!

**Comment by elizabethsjudd** [comment]
@asudoh I've updated the code based on your comments.

**Review by dakahn — APPROVED** [review]

File: packages/components/src/components/dropdown/dropdown.js [hunk 4]
```diff
@@ -91,53 +93,125 @@ class Dropdown extends mixin(
...
-      ? focused
-      : null;
+    let focusedNode;
+
+    // Using the latest semantic markup structure where trigger is a button
+    // @todo remove conditional once legacy structure is depreciated
+    if (this.element.querySelector(this.options.selectorTrigger)) {
+      const listNode = this.element.querySelector(this.options.selectorMenu);
+      const focusedId = listNode.getAttribute('aria-activedescendant');
+      focusedNode = focusedId ? listNode.querySelector(`#${focusedId}`) : null;
```
Comment (asudoh) [inline packages/components/src/components/dropdown/dropdown.js:197]: `getElementById()`?

**Review by asudoh — COMMENTED** [review]
Just one last comment in code. Would you be able to add test cases that represents the enhancement you did? Thanks @elizabethsjudd!

File: packages/components/src/components/dropdown/dropdown.js (hunk 4 above)
Comment (elizabethsjudd) [reply packages/components/src/components/dropdown/dropdown.js:197]: @asudoh `getElementById` only works from the `document` object which is why I opted for `querySelector` here.  I do believe that I could use `this.element.ownerDocument.getElementById` here though if you'd prefer that.

Comment (asudoh) [reply packages/components/src/components/dropdown/dropdown.js:197]: OK thank you for clarifying @elizabethsjudd! - Makes sense as-is.

**Review by asudoh — APPROVED** [review]
LGTM 👍 - Thanks @elizabethsjudd for all the updates and tests!

**Pull request merged**